etna/transforms/encoders/segment_encoder.py

```diff
@@ -44,12 +44,31 @@ def transform(self, df: pd.DataFrame) -> pd.DataFrame:
         -------
         :
             result dataframe
+
+        Raises
+        ------
+        ValueError:
+            If transform isn't fitted.
+        ValueError:
+            If there are segments that weren't present during training.
         """
-        encoded_matrix = self._le.transform(self._le.classes_)
-        encoded_matrix = encoded_matrix.reshape(len(self._le.classes_), -1).repeat(len(df), axis=1).T
+        segments = df.columns.get_level_values("segment").unique().tolist()
+
+        try:
+            new_segments = set(segments) - set(self._le.classes_)
+        except AttributeError:
+            raise ValueError("The transform isn't fitted!")
+
+        if len(new_segments) > 0:
+            raise ValueError(
+                f"This transform can't process segments that weren't present on train data: {new_segments}"
```

Comment:
Maybe we shouldn't include `new_segments` in error message. If one passes a completely different dataset with a large number of segments, the message will be cluttered.

Reply:
May be we should try: https://docs.python.org/3/library/reprlib.html